pkg/api/types.go

```diff
@@ -466,6 +470,30 @@ type GlusterfsVolumeSource struct {
 	ReadOnly bool `json:"readOnly,omitempty"`
 }
 
+// RBDVolumeSource represents a Rados Block Device Mount that lasts the lifetime of a pod
+type RBDVolumeSource struct {
+	// Required: EndpointsName is the endpoint name that represents a collection of Ceph monitors
+	EndpointsName string `json:"endpoints"`
+	// Required: RBDImage is the rados image name
+	RBDImage string `json:"image"`
+	// Required: Filesystem type to mount.
+	// Must be a filesystem type supported by the host operating system.
+	// Ex. "ext4", "xfs", "ntfs"
+	// TODO: how do we prevent errors in the filesystem from compromising the machine
+	FSType string `json:"fsType,omitempty"`
+	// Optional: RBDPool is the rados pool name,default is rbd
+	RBDPool string `json:"pool"`
+	// Optional: RadosUser is the rados user name, default is admin
+	RadosUser string `json:"user"`
+	// Optional: Keyring is the path to key ring for RBDUser, default is /etc/ceph/keyring
+	Keyring string `json:"keyring"`
+	// Optional: Secret is the authentication secret for RBDUser, default is empty.
+	Secret string `json:"secret" description:"secret is authentication secret. If provided, it overrides keyring"`
```

Comment:
> On Apr 15, 2015, at 5:58 PM, Tim Hockin <notifications@github.com> wrote:
> 
> In pkg/api/types.go:
> 
> > +	EndpointsName string `json:"endpoints"`
> > +	// Required: RBDImage is the rados image name
> > +	RBDImage string `json:"image"`
> > +	// Required: Filesystem type to mount.
> > +	// Must be a filesystem type supported by the host operating system.
> > +	// Ex. "ext4", "xfs", "ntfs"
> > +	// TODO: how do we prevent errors in the filesystem from compromising the machine
> > +	FSType string `json:"fsType,omitempty"`
> > +	// Optional: RBDPool is the rados pool name,default is rbd
> > +	RBDPool string `json:"pool"`
> > +	// Optional: RadosUser is the rados user name, default is admin
> > +	RadosUser string `json:"user"`
> > +	// Optional: Keyring is the path to key ring for RBDUser, default is /etc/ceph/keyring
> > +	Keyring string `json:"keyring"`
> > +	// Optional: Secret is the authentication secret for RBDUser, default is empty.
> > +	Secret string `json:"secret" description:"secret is authentication secret. If provided, it overrides keyring"`
> I'm not sure I know what "follow our existing secret conventions" means any more :)
> 
> Using ceph as a proxy for all such systems, there are two modes (Temporary) Volumes and PersistentVolumes. Temporary volumes are described as VolumeSource and PersistentVolumes as PersistentVolumeSource. Ceph as a PersistentVolume probably does not want to have per-user keys - users should not know it is ceph. Ceph as a (temporary) Volume might do fun things with keys, though the long-term future of this is open for debate.
> 
> I can think of a bunch of ways that this could work, but the simplest is to say ceph only exists as PersistentVolume and the secret it needs is a full objectref, or that it needs no keys at all. The user never sees or knows about the key, but the kubelet does.
> 
> Somewhat less simple is that ceph Volumes and PersistentVolumes need a Secret. Plain Volumes must fetch the secret from in-namespace.  PersistentVolumes exist above namespaces and as such can fetch a secret from any namespace.
> 
This last option is to me the right choice, because a PV and VS are not the same object.  One is about hiding details from the user, the other is about exercising the flow directly.  I care less about VS in general than PV, but I don't see PV and VS as requiring equivalency, and to remove the ability to use ceph as VS (can we verb that noun?) seems premature.

Why not say "Ceph VS" can do everything a "Secret VS" can do?  (Reference a secret in current namespace by name)
> —
> Reply to this email directly or view it on GitHub.
>